### ISSUE TITLE
U2F Token

### ISSUE THREAD

**Issue by cornelinux** [privacyidea/privacyidea#182]
U2F hardware tokens are used by end users to authenticate anonymously/pseudonymously to correspondingly enabled websites.
U2F is a system that is designed to work without a central management. 
But it provides cheap hardware tokens.
Unfortunately these hardware tokens all contain a preseeded secret key, which could enable the vendors to identify the "anonymous" users.

We would prefer hardware tokens, that can be initialized. 
Nevertheless due to the pricing of those tokens, it might be interesting to support U2F tokens this way, that a user can register his U2F token and use it to authenticate against privacyIDEA.

**Note**: In this case again the user would loose his anonymity.

https://fidoalliance.org/specifications/download/

Take a look at:
https://github.com/Yubico/u2fval

**Comment by jh23453** [comment]
LinOTP has an implementation in https://github.com/LinOTP/LinOTP/blob/master/linotpd/src/linotp/lib/tokens/u2ftoken.py and the related commits.

**Comment by cornelinux** [comment]
Thanks a lot for the heads up.

Thats true. Well, I looked at it and with 849 sloc it look rather expensive to me.
There was a similar case with ORCA. Granted, the ocra2token.py with ~1100 sloc of LinOTP might be a more complete implementation, but I rather took the chance to implement it a new in privacyIDEA's tiqrtoken.py with ~420 sloc. I also looked at the python oath module - but integrating the oath module which was not packed for debian, seemed more complicated to me.

And there are also some significant changes. While LinOTP introduced new controllers / endpoints for OCRA and U2F, privacyIDEA tries to avoid this by using the ttype endpoint. (see http://privacyidea.readthedocs.org/en/latest/modules/api/ttype.html) I think it is a rather good idea, to give any token the possibility to define some token specific endpoint action. Avoiding adding new endpoints with new token types.

So I guess I will take a look at some implementations to understand the specification but implement it anew.
